README.TXT
DOSMid Copyright (C) 2014-2018 Mateusz Viste
Creative Music System / Game Blaster output support version (C) Tronix 2021

Used piece of software codes from third-party peoples, eg:
- DEATHSHADOW'S Paku-Paku game by Jason M. Knight
- SAATunes v 1.04 lib by Bobcatmodder Jacob Field
- Arduino YM2149 MIDI Synth by yukimizake

New DOSMid options:

 /cms=XXX  Force dosmid to use Creative Music System / Game Blaster
           on port XXX. The port part is optional, that means you can 
           use "/cms" to just force CMS usage on default 220h port.


============================= ORIGINAL DOSMID.TXT
           
         DOSMid - a low-requirements MIDI and MUS player for DOS
                    http://dosmid.sourceforge.net


*** INTRO ***

DOSMid is a MIDI, MUS and RMID player for DOS. It's a real mode application
designed to run on very modest hardware configurations. It plays standard MIDI
files (both format 0 and format 1), as well as MIDI in RIFF (ie. RMID), and
even MUS tunes (as used by Id Software in numerous games like Doom, Heretic,
Hexen, etc).
DOSMid requires MIDI-capable hardware of some sort. Note, that some sound
cards have an MPU-401 interface, although many need an additional 'wavetable' daughterboard to produce actual MIDI sound.
DOSMid can also emulate MIDI through FM synthesis using an OPL chip (that is
one of the Yamaha YM3812 or YMF262 chips, found on most sound cards from the
nineties) - be warned however that, most of the time, such MIDI-over-OPL
emulation will yeld less than desirable results, unless the MIDI file was
specifically crafted for OPL.

Minimum requirements:
 - a compatible synthesizer (wavetable, OPL or external - see the compat list)
 - an 8086-compatible CPU
 - 100K of available conventional memory and some XMS
 - 250K or so available conventional memory if you do not have XMS
 - a video card able to display an 80x25 monochrome text mode

Highly recommended:
 - 80286 CPU for a guaranteed lag-free experience even on complex MIDI files
 - VGA graphic with a color monitor
 - 512K of available XMS memory


*** USAGE ***

During runtime, DOSMid can be controlled with the keyboard:
 ESC       Quits to DOS
 +/-       Volume up/down
 SPACE     Pause the song (press any key to resume)
 ENTER     Skip to next song of the playlist
 BKSPC     Jump to previous song of the playlist (doesn't work with /random)

DOSMid accepts several command-line options, as listed below:

DOSMID [options] file.mid (or m3u playlist)

 /noxms    Use conventional memory instead of XMS. This is obviously useful
           only if you don't have XMS. Don't use this option otherwise, since
           without XMS you won't be able to load big MIDI files.
 /xmsdelay Wait 2ms before each XMS access. Such waiting is required sometimes
           when the MPU controller is emulated by a TSR driver (specifically,
           the AWEUTIL driver used with SoundBlaster AWE 32/64 cards happens
           to crash if XMS accesses are not slightly delayed).
 /mpu=XXX  Force dosmid to use MPU-401 on port XXX. If not forced, DOSMID
           scans the BLASTER environment variable for the MPU port, and if not
           found, it fallbacks to port 330h. The port part is optional, that
           means you can use "/mpu" to just force MPU usage.
 /awe=XXX  Use the EMU8000 synth chip found on SoundBlaster AWE32/AWE64 cards
           on port XXX (the port is optional, you can specify just "/awe").
           KNOWN BUG: On some AWE cards, the FM music module becomes muted or
           noisy after using the EMU8000 chip. This is not a bug in DOSMid,
           and happens with other applications using AWE as well. If you have
           this problem, execute AWEUTIL /S after using DOSMid to reinit FM.
           I observed this problem on an AWE64 CT4390 ("Gold"), but not on an
           AWE32 CT2760.
 /opl=XXX  Use an OPL2-compatible chip on port XXX. This should be a last
           resort option if you don't have any wavetable device. Do NOT expect
           pleasing results. The port part is optional ("/opl" will default to
           port 388h).
 /sbmidi=XXX Drives an external synth connected to the gameport of your Sound
           Blaster card. The port part is optional ("/sbmidi" will use the
           port read from BLASTER, or fallback to 220h).
 /com=XXX  Send MIDI messages out via the RS232 port at I/O port XXX. This can
  or /com1 be used to hook a hardware synth to a computer with no MIDI
  or /com2 interface, only a standard serial port. DOSMid does NOT reconfigure
  or /com3 the COM port, so you should take care of setting it correctly, for
  or /com4 example using the 'MODE COM1: ...' command. The port part is not
           optional, you are expected to pass the hexadecimal I/O address of
           the RS232 port you wish to use (for example "/com=3f8" is pointing
           to COM1 on most BIOS implementations).
           It is also possible to use simpler "/com1", "/com2", "/com3" and
           "/com4" switches. These will autodetect the correct I/O port.
 /gus      Use the Gravis UltraSound card, relying on its ULTRAMID API.
 /syx=FILE Uses SYSEX instructions stored in FILE for initializing the MIDI
           device. FILE must be in "SYX" format, and can contain one or more
           SYSEX messages.
 /delay=X  Insert an extra delay of X ms (X being in the range 1..9000) before
           playing the MIDI file. Setting this to 100 or 200 might help in
           some cases where the sound hardware needs more time to initialize
           completely, or when complex sysex data is fed via the /syx option.
           You can also use this simply to make the silence longer between
           the files of your playlist.
 /sbnk=FILE Makes DOSMid load a custom sound bank. This is supported only on
           OPL and AWE hardware. The sound bank file must be in the IBK format
           when used with OPL, and SBK format for AWE. OPL accepts one or two
           IBK files (eg. /sbnk=file1.ibk,file2.ibk). If two are provided,
           then the first one will be used for the standard 128-instruments GM
           set, and the second one for defining percussion instruments.
 /log=FILE Logs all DOSMid activity to FILE. This is a debugging feature that
           you shouldn't be interested in. Beware, the log file can get pretty
           big (MUCH bigger than the MIDI file you are playing).
 /fullcpu  Do not let DOSMid being CPU-friendly. By default DOSMid issues an
           INT 28h when idle, to let the system be gentler on the CPU, but on
           some hardware this might lead to degraded sound performance.
 /dontstop Never ask the user to press a key after an error occurs. This is
           useful if you want to play a long playlist and don't care about
           bad MIDI files, simply skipping them (or if you play a single file
           and wish that DOSMid exit immediately if the file is unplayable).
 /random   randomize playlist order
 /nosound  Disable sound (not very useful for a music player!)

Note: All the above options can also be written to the DOSMID.CFG file. This
file is loaded by DOSMID and used as default parameters that can still be
overloaded by command-line options.


*** THE BLASTER VARIABLE ***

When not forced into a specific configuration via command-line switches,
DOSMid scans the BLASTER environment variable to find out the most desirable
settings. A BLASTER environment variable usually looks similar to this:

  SET BLASTER=A220 I5 D1 T3 P330 H6 E620

The bits DOSMid is interested in are "A220", "P330" and "E620". P330 provides
the port address of the MPU-401 MIDI interface, while E620 tells the port
address of the EMU8000 onboard synth (available only on 'AWE' models). A220,
on the other hand, provides the base I/O address of the SoundBlaster card, so
DOSMid can output directly to the card's MIDI port. If not instructed
otherwise, DOSMid will always try to use the EMU8000 synth if found in the
BLASTER string, and if not, it will use the MPU-401. If neither of them are
found in the BLASTER string, or if there is no BLASTER variable at all, then
DOSMid will try to fall back to FM synthesis on port 388h, unless an ULTRADIR
environment variable is present, in which case it will try to detect a GUS
through the ULTRAMID TSR API.


*** COMPATIBLE SOUND HARDWARE ***

DOSMid supports a variety of MIDI hardware. Here below I list the types of
hardware that DOSMid is able to talk to:

External MIDI synthesizers:
 - Connected through a MIDI/game port: either using an industry standard
   MPU-401 interface, or SoundBlaster MIDI port,
 - Connected through a "COM" (RS232) port: many synthesizers come with an
   RS-232 port that can be used instead of the standard MIDI port. This is
   supported by Roland SoundCanvas models, Yamaha PSR series, the Miracle
   Piano, many Korg devices... You could even use a software solution based
   on ttymidi...

Internal MIDI synthesizers:
 - Available through a virtual MPU-401 interface: some AzTech Waverider 32
   models, some versions of the HighScreen SoundBoostar 16, SoundBlaster 64
   cards using the 'AWEUTIL' MPU emulator,
 - Based on the EMU8000 chip (SoundBlaster 32, SoundBlaster 64),
 - Gravis UltraSound (GUS) cards, through their ULTRAMID driver,
 - Based on an OPL2 or OPL3-compatible chip (most ISA sound cards from the
   nineties: Adlib, all SoundBlaster ISA models, Opti, AzTech...).


*** BUILDING ***

DOSMid is compiled with OpenWatcom 1.9. The entire build process is automated
through a Makefile file, so if you wish to rebuild DOSMid, all you have to do
is type "wmake". A few compile-time options are available to disable features
you might not want: edit the Makefile and adapt FEATURES to your likeness. The
FEATURES list is documented in the Makefile. The default build comes with all
features compiled in, so you really should fiddle with FEATURES only in case
you desperately need to lower DOSMid's memory footprint.


*** CONTACT ***

If you enjoy DOSMid, or noticed any bugs, I'd love to hear about it! You won't
see my e-mail address here, but you will surely find some contact pointers on
my home page: http://mateusz.viste.fr


*** LICENSE ***

Copyright (C) 2014-2018 Mateusz Viste
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


HISTORY.TXT

* DOSMid history file *


DOSMid v0.9.5 [10 Dec 2018]
 - Fixed file-reading module (buff corruption on first cache boundary read).

DOSMid v0.9.4 [09 Dec 2018]
 - UI displays channels as 'piano' if no instrument is explicitely loaded,
 - reduced libc dependencies (smaller footprint),
 - displaying the MIDI tracks count,
 - displaying the amount of memory used by the MIDI file,
 - backspace key jumps to previous song of the playlist,
 - minor memory optimizations,
 - fixed a memory leak in midi code,
 - fixed RIFF-RMID header decoding,
 - fixed playlist gap delay computation (fixed /delay behavior, too),
 - adaptations to compile in small memory model (smaller, faster),
 - removed /log from the default build,
 - /noxms can be passed via config file now,
 - GUS support (requires the ULTRAMID TSR).

DOSMid v0.9.3 [06 Feb 2018]
 - memory optimization (25% less RAM needed when loading files),
 - improved memory usage when using /noxms.

DOSMid v0.9.2 [06 Jun 2017]
 - fixed parsing of the /SBMIDI=xxx value (bug reported by James-F),
 - sequential playlist playing (inspired by a patch from Graham Wiseman).

DOSMid v0.9.1 [19 Nov 2016]
 - resetting channels to "all piano" before playing MIDI files,
 - syx files can be set from within DOSMID.CFG now,
 - configurable extra delay before playing files (/delay=X).

DOSMid v0.9 [21 May 2016]
 - increased SYX delaying to 40ms (solves buffering troubles on MT-32 gears),
 - fixes around how OPL percussion instruments are emulated,
 - support for custom sound banks (IBK on OPL, SBK on AWE),
 - a configuration file allows to set default settings (dosmid.cfg),
 - the 2s silence gap before songs is inserted only in playlist mode now,
 - sound hardware is initialized once, instead of being inited for every song,
 - fixed playlist parsing so DOSMid won't freeze on an empty m3u playlist,
 - load more textual information from the MIDI file (titles, markers, text),
 - file's information is displayed on a dynamic scroll when over 5 lines,
 - reorganized the display so more MIDI textual data fits on the screen,
 - fixed OPL detection for cards intolerant to hw calls during init (CT2760),
 - command-line arguments are case-insensitive,
 - renamed the /delay command-line option to /xmsdelay,
 - DOSMid refuses to play corrupted files, instead of ignoring MIDI errors.

DOSMid v0.8 [18 Sep 2015]
 - SBMIDI support (can drive MIDI synths attached to SB 1.x and 2.x cards),
 - serial (RS-232) output support (/com=xxx or /com1..4),
 - implemented 'controller' MIDI support on OPL output,
 - added support for relaying SYSEX messages to the MIDI device,
 - DOSMid adds a 2s silence gap before each song,
 - support for monochrome video modes (autodetected),
 - fixed parsing of hex I/O ports with non-numerical chars (like "3f8"),
 - reduced timing inaccuracy from ~0.02% to <0.005% (relies on PIT's quality),
 - restoring initial video mode after DOSMid quits,
 - reliable detection and reporting of out of memory conditions,
 - /noxms handles more than 64K of memory (bigger files can be played),
 - support for custom SYSEX initialization via a SYX file (/syx=FILE),
 - "expensive" features can be easily disabled at compile-time (AWE, OPL),
 - screen refresh optimizations - runs without lagging on my 8086 now.

DOSMid v0.7 [29 Jul 2015]
 - native support for SB AWE32/AWE64 sound cards (no need for AWEUTIL),
 - OPL2/OPL3 output support via emulated (waveform) GM patches,
 - support for *.MUS files, as used in many Doom-like games by Id Software,
 - optimized MIDI loading routines - MIDI files are loading twice as fast now,
 - added a pause function (space key),
 - song's filename is displayed during load time,
 - the /nodelay option is no longer default (use /delay to force).

DOSMid v0.6.1 [01 Jul 2015]
 - fixed a NULL pointer dereferencing bug,
 - possible memory access troubles are reported clearly to the user,
 - forcing GM patches to MPU because some songs rely on default instruments,
 - refactored the sound output code (and added /nosound).

DOSMid v0.6 [22 Jun 2015]
 - fixed a timing issue leading sometimes to desynchronization of tracks,
 - fixed a bug that could potentially lead to XMS memory trashing,
 - configurable MPU port via the BLASTER variable or /mpu=XXX,
 - DOSMid fails gracefully when no MPU401 is present (instead of freezing),
 - added software volume controls under '+' and '-' keys,
 - the MPU is reinitialized to default values before playing any new tracks,
 - interface fitted with a sexier color scheme,
 - using CPU-friendly calls when waiting for notes (disabled with /fullcpu),
 - very basic support for reading M3U playlists,
 - added a verbose mode for debugging needs (/log=FILE).

DOSMid v0.5 [10 Aug 2014]
 - first public release.


GM.H
/*
 * This file contains GM instruments names, as displayed by DOSMid.
 * Copyright (C) 2014 Mateusz Viste
 */

static char *gmset[] = {
/* PIANO                           CHROMATIC PERCUSSION
1    Acoustic Grand             9   Celesta
2    Bright Acoustic           10   Glockenspiel
3    Electric Grand            11   Music Box
4    Honky-Tonk                12   Vibraphone
5    Electric Piano 1          13   Marimba
6    Electric Piano 2          14   Xylophone
7    Harpsichord               15   Tubular Bells
8    Clavinet                  16   Dulcimer */

"Acoustic Grand",
"Bright Acoustic",
"Electric Grand",
"Honky-Tonk",
"Electric Piano",
"Electric Piano",
"Harpsichord",
"Clavinet",
"Celesta",
"Glockenspiel",
"Music Box",
"Vibraphone",
"Marimba",
"Xylophone",
"Tubular Bells",
"Dulcimer",

/* ORGAN                          GUITAR
17   Drawbar Organ             25   Nylon String Guitar
18   Percussive Organ          26   Steel String Guitar
19   Rock Organ                27   Electric Jazz Guitar
20   Church Organ              28   Electric Clean Guitar
21   Reed Organ                29   Electric Muted Guitar
22   Accoridan                 30   Overdriven Guitar
23   Harmonica                 31   Distortion Guitar
24   Tango Accordian           32   Guitar Harmonics */

"Drawbar Organ",
"Percuss. Organ",
"Rock Organ",
"Church Organ",
"Reed Organ",
"Accordian",
"Harmonica",
"Tango Accordian",
"Nylon Guitar",
"Steel Guitar",
"E. Jazz Guitar",
"E. Clean Guitar",
"E. Muted Guitar",
"Overdr. Guitar",
"Distort. Guitar",
"Guitar Harmon.",

/* BASS                           SOLO STRINGS
33   Acoustic Bass             41   Violin
34   Electric Bass(finger)     42   Viola
35   Electric Bass(pick)       43   Cello
36   Fretless Bass             44   Contrabass
37   Slap Bass 1               45   Tremolo Strings
38   Slap Bass 2               46   Pizzicato Strings
39   Synth Bass 1              47   Orchestral Strings
40   Synth Bass 2              48   Timpani */

"Acoustic Bass",
"Electric Bass",
"Electric Bass",
"Fretless Bass",
"Slap Bass 1",
"Slap Bass 2",
"Synth Bass 1",
"Synth Bass 2",
"Violin",
"Viola",
"Cello",
"Contrabass",
"Tremolo Strings",
"Pizzicato Str.",
"Orchestral Str.",
"Timpani",

/* ENSEMBLE                       BRASS
49   String Ensemble 1         57   Trumpet
50   String Ensemble 2         58   Trombone
51   SynthStrings 1            59   Tuba
52   SynthStrings 2            60   Muted Trumpet
53   Choir Aahs                61   French Horn
54   Voice Oohs                62   Brass Section
55   Synth Voice               63   SynthBrass 1
56   Orchestra Hit             64   SynthBrass 2 */

"String Ensemble",
"String Ensemble",
"Synth Strings",
"Synth Strings",
"Choir Aahs",
"Voice Oohs",
"Synth Voice",
"Orchestra Hit",
"Trumpet",
"Trombone",
"Tuba",
"Muted Trumpet",
"French Horn",
"Brass Section",
"Synth Brass 1",
"Synth Brass 2",

/* REED                           PIPE
65   Soprano Sax               73   Piccolo
66   Alto Sax                  74   Flute
67   Tenor Sax                 75   Recorder
68   Baritone Sax              76   Pan Flute
69   Oboe                      77   Blown Bottle
70   English Horn              78   Skakuhachi
71   Bassoon                   79   Whistle
72   Clarinet                  80   Ocarina */

"Soprano Sax",
"Alto Sax",
"Tenor Sax",
"Baritone Sax",
"Oboe",
"English Horn",
"Bassoon",
"Clarinet",
"Piccolo",
"Flute",
"Recorder",
"Pan Flute",
"Blown Bottle",
"Skakuhachi",
"Whistle",
"Ocarina",

/* SYNTH LEAD                     SYNTH PAD
81   Lead 1 (square)           89   Pad 1 (new age)
82   Lead 2 (sawtooth)         90   Pad 2 (warm)
83   Lead 3 (calliope)         91   Pad 3 (polysynth)
84   Lead 4 (chiff)            92   Pad 4 (choir)
85   Lead 5 (charang)          93   Pad 5 (bowed)
86   Lead 6 (voice)            94   Pad 6 (metallic)
87   Lead 7 (fifths)           95   Pad 7 (halo)
88   Lead 8 (bass+lead)        96   Pad 8 (sweep) */

"Square Synth",
"Sawtooth Synth",
"Calliope Synth",
"Chiff Synth",
"Charang Synth",
"Voice Synth",
"Fifths Synth",
"Bass+Lead Synth",
"New Age Synth",
"Warm Synth",
"Polysynth",
"Choir Synth",
"Bowed Synth",
"Metallic Synth",
"Halo Synth",
"Sweep Synth",

/* SYNTH EFFECTS                  ETHNIC
 97  FX 1 (rain)               105   Sitar
 98  FX 2 (soundtrack)         106   Banjo
 99  FX 3 (crystal)            107   Shamisen
100  FX 4 (atmosphere)         108   Koto
101  FX 5 (brightness)         109   Kalimba
102  FX 6 (goblins)            110   Bagpipe
103  FX 7 (echoes)             111   Fiddle
104  FX 8 (sci-fi)             112   Shanai */

"Rain (FX)",
"Soundtrack (FX)",
"Crystal (FX)",
"Atmosphere (FX)",
"Brightness (FX)",
"Goblins (FX)",
"Echoes (FX)",
"Sci-Fi (FX)",
"Sitar",
"Banjo",
"Shamisen",
"Koto",
"Kalimba",
"Bagpipe",
"Fiddle",
"Shanai",

/* PERCUSSIVE                     SOUND EFFECTS
113  Tinkle Bell               121   Guitar Fret Noise
114  Agogo                     122   Breath Noise
115  Steel Drums               123   Seashore
116  Woodblock                 124   Bird Tweet
117  Taiko Drum                125   Telephone Ring
118  Melodic Tom               126   Helicopter
119  Synth Drum                127   Applause
120  Reverse Cymbal            128   Gunshot */

"Tinkle Bell",
"Agogo",
"Steel Drums",
"Woodblock",
"Taiko Drum",
"Melodic Tom",
"Synth Drum",
"Reverse Cymbal",
"Guitar Noise",
"Breath Noise",
"Seashore",
"Bird Tweet",
"Telephone Ring",
"Helicopter",
"Applause",
"Gunshot"

};
